tailssh: CI flake in TestSSHRecordingCancelsSessionsOnUploadFailure

```
--- FAIL: TestSSHRecordingCancelsSessionsOnUploadFailure (0.11s)
    tailssh.go:255: ssh-conn-20230327T232538-f1e060e10c: handling conn: 100.100.100.101:2231->alice@100.100.100.102:22
    tailssh.go:255: ssh-conn-20230327T232538-f1e060e10c: starting session: sess-20230327T232538-5bd83278d2
    logger.go:98: ssh-session(sess-20230327T232538-5bd83278d2): handling new SSH connection from peer (100.100.100.101) to ssh-user "runner"
    logger.go:98: ssh-session(sess-20230327T232538-5bd83278d2): access granted to peer as ssh-user "runner"
    logger.go:98: ssh-session(sess-20230327T232538-5bd83278d2): starting asciinema recording to 127.0.0.1:36829
    logger.go:98: ssh-session(sess-20230327T232538-5bd83278d2): recording: server responded with 403 Forbidden
    logger.go:98: ssh-session(sess-20230327T232538-5bd83278d2): startNewRecording: io: read/write on closed pipe
    --- FAIL: TestSSHRecordingCancelsSessionsOnUploadFailure/upload-denied (0.00s)
        tailssh_test.go:412: client established session
        tailssh_test.go:415: client got: "can't start new recording\r\n": Process exited with status 1
        tailssh_test.go:420: client got "can't start new recording\r\n", want "recording: server responded with 403 Forbidden\r\n"
    tailssh.go:255: ssh-conn-20230327T232538-f3486f3b09: handling conn: 100.100.100.101:2231->alice@100.100.100.102:22
    tailssh.go:255: ssh-conn-20230327T232538-f3486f3b09: starting session: sess-20230327T232538-2af71b3dc3
    logger.go:98: ssh-session(sess-20230327T232538-2af71b3dc3): handling new SSH connection from peer (100.100.100.101) to ssh-user "runner"
    logger.go:98: ssh-session(sess-20230327T232538-2af71b3dc3): access granted to peer as ssh-user "runner"
    logger.go:98: ssh-session(sess-20230327T232538-2af71b3dc3): starting asciinema recording to 127.0.0.1:36[829](https://github.com/tailscale/tailscale/actions/runs/4537255517/jobs/7995449008?pr=7705#step:7:830)
    logger.go:98: ssh-session(sess-20230327T232538-2af71b3dc3): starting non-pty command: [/bin/bash -c echo hello && sleep 1 && echo world]
    logger.go:98: ssh-session(sess-20230327T232538-2af71b3dc3): recording: server responded with 500 Internal Server Error
    logger.go:98: ssh-session(sess-20230327T232538-2af71b3dc3): terminating SSH session from 100.100.100.101: recording: server responded with 500 Internal Server Error
    logger.go:98: ssh-session(sess-20230327T232538-2af71b3dc3): stdout copy: read |0: file already closed
    logger.go:98: ssh-session(sess-20230327T232538-2af71b3dc3): stderr copy: read |0: file already closed
    logger.go:98: ssh-session(sess-20230327T232538-2af71b3dc3): Wait: code=-1
FAIL
exit status 1
FAIL	tailscale.com/ssh/tailssh	0.677s
```

https://github.com/tailscale/tailscale/actions/runs/4537255517/jobs/7995449008?pr=7705

cc @maisem

--- a/ssh/tailssh/tailssh.go
+++ b/ssh/tailssh/tailssh.go
@@ -1065,7 +1065,12 @@ func (ss *sshSession) run() {
 			var err error
 			rec, err = ss.startNewRecording()
 			if err != nil {
-				fmt.Fprintf(ss, "can't start new recording\r\n")
+				var uve userVisibleError
+				if errors.As(err, &uve) {
+					fmt.Fprintf(ss, "%s\r\n", uve)
+				} else {
+					fmt.Fprintf(ss, "can't start new recording\r\n")
+				}
 				ss.logf("startNewRecording: %v", err)
 				ss.Exit(1)
 				return
@@ -1079,7 +1084,7 @@ func (ss *sshSession) run() {
 		logf("start failed: %v", err.Error())
 		if errors.Is(err, context.Canceled) {
 			err := context.Cause(ss.ctx)
-			uve := userVisibleError{}
+			var uve userVisibleError
 			if errors.As(err, &uve) {
 				fmt.Fprintf(ss, "%s\r\n", uve)
 			}
@@ -1500,6 +1505,12 @@ func (ss *sshSession) startNewRecording() (_ *recording, err error) {
 	}
 	j = append(j, '\n')
 	if _, err := pw.Write(j); err != nil {
+		if errors.Is(err, io.ErrClosedPipe) && ss.ctx.Err() != nil {
+			// If we got an io.ErrClosedPipe, it's likely because
+			// the recording server closed the connection on us. Return
+			// the original context error instead.
+			return nil, context.Cause(ss.ctx)
+		}
 		return nil, err
 	}
 	return rec, nil

